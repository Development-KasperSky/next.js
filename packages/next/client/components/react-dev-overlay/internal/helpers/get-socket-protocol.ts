export function getSocketProtocol(assetPrefix: string): string {
  let protocol = window.location.protocol

  try {
    // assetPrefix is a url
    protocol = new URL(assetPrefix).protocol
  } catch (_) {}

  return protocol === 'http:' ? 'ws' : 'wss'
}
