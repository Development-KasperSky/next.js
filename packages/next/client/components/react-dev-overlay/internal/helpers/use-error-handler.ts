import { useEffect, useRef } from 'react'

export function useErrorHandler(
  handleOnUnhandledError: (event: WindowEventMap['error']) => void,
  handleOnUnhandledRejection: (
    event: WindowEventMap['unhandledrejection']
  ) => void
) {
  const stacktraceLimitRef = useRef<undefined | number>()

  useEffect(() => {
    try {
      const limit = Error.stackTraceLimit
      Error.stackTraceLimit = 50
      stacktraceLimitRef.current = limit
    } catch {}

    window.addEventListener('error', handleOnUnhandledError)
    window.addEventListener('unhandledrejection', handleOnUnhandledRejection)
    return () => {
      if (stacktraceLimitRef.current !== undefined) {
        try {
          Error.stackTraceLimit = stacktraceLimitRef.current
        } catch {}
        stacktraceLimitRef.current = undefined
      }

      window.removeEventListener('error', handleOnUnhandledError)
      window.removeEventListener(
        'unhandledrejection',
        handleOnUnhandledRejection
      )
    }
  }, [handleOnUnhandledError, handleOnUnhandledRejection])
}
