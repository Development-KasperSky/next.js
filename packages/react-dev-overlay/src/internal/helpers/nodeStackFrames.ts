import { parse, StackFrame } from 'stacktrace-parser'

export function getFilesystemFrame(frame: StackFrame): StackFrame {
  const f: StackFrame = { ...frame }

  if (typeof f.file === 'string') {
    if (
      // Posix:
      f.file.startsWith('/') ||
      // Win32:
      /^[a-z]:\\/i.test(f.file) ||
      // Win32 UNC:
      f.file.startsWith('\\\\')
    ) {
      f.file = `file://${f.file}`
    }
  }

  return f
}

const symbolError = Symbol('NextjsError')

export function getErrorSource(error: Error): 'server' | 'edge-server' | null {
  return (error as any)[symbolError] || null
}

export function getServerError(
  error: Error,
  type: 'edge-server' | 'server'
): Error {
  let n: Error
  try {
    throw new Error(error.message)
  } catch (e) {
    n = e as Error
  }

  n.name = error.name
  try {
    n.stack = parse(error.stack!)
      .map(getFilesystemFrame)
      .map((f) => {
        let str = `    at ${f.methodName}`
        if (f.file) {
          let loc = f.file
          if (f.lineNumber) {
            loc += `:${f.lineNumber}`
            if (f.column) {
              loc += `:${f.column}`
            }
          }
          str += ` (${loc})`
        }
        return str
      })
      .join('\n')
  } catch {
    n.stack = error.stack
  }

  Object.defineProperty(n, symbolError, {
    writable: false,
    enumerable: false,
    configurable: false,
    value: type,
  })
  return n
}
