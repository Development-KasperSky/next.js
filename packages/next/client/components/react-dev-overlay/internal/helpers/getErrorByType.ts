import {
  ACTION_UNHANDLED_ERROR,
  ACTION_UNHANDLED_REJECTION,
} from '../error-overlay-reducer'
import { SupportedErrorEvent } from '../container/Errors'
import { getErrorSource } from './nodeStackFrames'
import { getOriginalStackFrames, OriginalStackFrame } from './stack-frame'

export type ReadyRuntimeError = {
  id: number
  runtime: true
  error: Error
  frames: OriginalStackFrame[]
}

export async function getErrorByType(
  ev: SupportedErrorEvent
): Promise<ReadyRuntimeError> {
  const { id, event } = ev
  switch (event.type) {
    case ACTION_UNHANDLED_ERROR:
    case ACTION_UNHANDLED_REJECTION: {
      return {
        id,
        runtime: true,
        error: event.reason,
        frames: await getOriginalStackFrames(
          event.frames,
          getErrorSource(event.reason),
          event.reason.toString()
        ),
      }
    }
    default: {
      break
    }
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const _: never = event
  throw new Error('type system invariant violation')
}
