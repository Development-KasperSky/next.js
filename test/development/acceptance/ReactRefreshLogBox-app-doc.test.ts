import { sandbox } from './helpers'
import { createNext } from 'e2e-utils'
import { NextInstance } from 'test/lib/next-modes/base'

describe('ReactRefreshLogBox', () => {
  let next: NextInstance

  beforeAll(async () => {
    next = await createNext({
      files: {},
      skipStart: true,
    })
  })
  afterAll(() => next.destroy())

  test('empty _app shows logbox', async () => {
    const { session, cleanup } = await sandbox(
      next,
      new Map([
        [
          'pages/_app.js',
          `
            
          `,
        ],
      ])
    )
    expect(await session.hasRedbox(true)).toBe(true)
    expect(await session.getRedboxDescription()).toMatchInlineSnapshot(
      `"Error: The default export is not a React Component in page: \\"/_app\\""`
    )

    await session.patch(
      'pages/_app.js',
      `
        function MyApp({ Component, pageProps }) {
          return <Component {...pageProps} />;
        }
        export default MyApp
      `
    )
    expect(await session.hasRedbox()).toBe(false)
    await cleanup()
  })

  test('empty _document shows logbox', async () => {
    const { session, cleanup } = await sandbox(
      next,
      new Map([
        [
          'pages/_document.js',
          `
            
          `,
        ],
      ])
    )
    expect(await session.hasRedbox(true)).toBe(true)
    expect(await session.getRedboxDescription()).toMatchInlineSnapshot(
      `"Error: The default export is not a React Component in page: \\"/_document\\""`
    )

    await session.patch(
      'pages/_document.js',
      `
        import Document, { Html, Head, Main, NextScript } from 'next/document'

        class MyDocument extends Document {
          static async getInitialProps(ctx) {
            const initialProps = await Document.getInitialProps(ctx)
            return { ...initialProps }
          }

          render() {
            return (
              <Html>
                <Head />
                <body>
                  <Main />
                  <NextScript />
                </body>
              </Html>
            )
          }
        }

        export default MyDocument
      `
    )
    expect(await session.hasRedbox()).toBe(false)
    await cleanup()
  })

  test('_app syntax error shows logbox', async () => {
    const { session, cleanup } = await sandbox(
      next,
      new Map([
        [
          'pages/_app.js',
          `
            function MyApp({ Component, pageProps }) {
              return <<Component {...pageProps} />;
            }
            export default MyApp
          `,
        ],
      ])
    )
    expect(await session.hasRedbox(true)).toBe(true)
    expect(await session.getRedboxSource()).toMatchSnapshot()

    await session.patch(
      'pages/_app.js',
      `
        function MyApp({ Component, pageProps }) {
          return <Component {...pageProps} />;
        }
        export default MyApp
      `
    )
    expect(await session.hasRedbox()).toBe(false)
    await cleanup()
  })

  test('_document syntax error shows logbox', async () => {
    const { session, cleanup } = await sandbox(
      next,
      new Map([
        [
          'pages/_document.js',
          `
            import Document, { Html, Head, Main, NextScript } from 'next/document'

            class MyDocument extends Document {{
              static async getInitialProps(ctx) {
                const initialProps = await Document.getInitialProps(ctx)
                return { ...initialProps }
              }

              render() {
                return (
                  <Html>
                    <Head />
                    <body>
                      <Main />
                      <NextScript />
                    </body>
                  </Html>
                )
              }
            }

            export default MyDocument
          `,
        ],
      ])
    )
    expect(await session.hasRedbox(true)).toBe(true)
    expect(await session.getRedboxSource()).toMatchSnapshot()

    await session.patch(
      'pages/_document.js',
      `
        import Document, { Html, Head, Main, NextScript } from 'next/document'

        class MyDocument extends Document {
          static async getInitialProps(ctx) {
            const initialProps = await Document.getInitialProps(ctx)
            return { ...initialProps }
          }

          render() {
            return (
              <Html>
                <Head />
                <body>
                  <Main />
                  <NextScript />
                </body>
              </Html>
            )
          }
        }

        export default MyDocument
      `
    )
    expect(await session.hasRedbox()).toBe(false)
    await cleanup()
  })
})
