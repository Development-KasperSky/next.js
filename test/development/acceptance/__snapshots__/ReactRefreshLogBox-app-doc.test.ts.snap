// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactRefreshLogBox _app syntax error shows logbox 1`] = `
"./pages/_app.js
Error: error: Expression expected
 
  |
3 |               return <<Component {...pageProps} />;
  |                       ^

error: Unexpected token \`jsx name (Component)\`. Expected this, import, async, function, [ for array literal, { for object literal, @ for decorator, function, class, null, true, false, number, bigint, string, regexp, \` for template literal, (, or an identifier
 
  |
3 |               return <<Component {...pageProps} />;
  |                        ^^^^^^^^^

Caused by:
    0: failed to process js file
    1: Syntax Error"
`;

exports[`ReactRefreshLogBox _document syntax error shows logbox 1`] = `
"./pages/_document.js
Error: error: Unexpected token \`{\`. Expected identifier, string literal, numeric literal or [ for the computed key
 
  |
4 |             class MyDocument extends Document {{
  |                                                ^

Caused by:
    0: failed to process js file
    1: Syntax Error"
`;
